### PR TITLE
REF: make Grouping less stateful

### PR DESCRIPTION
**Pull request by jbrockmendel**
This gets rid of _make_codes, which sets two state variables, in favor of a cache_readonly _codes_and_uniques

### CODE REVIEW

File: pandas/core/groupby/grouper.py [hunk 1]
```diff
@@ -587,20 +588,23 @@ def indices(self):
 
     @property
     def codes(self) -> np.ndarray:
-        if self._passed_categorical:
-            # we make a CategoricalIndex out of the cat grouper
-            # preserving the categories / ordered attributes
-            cat = self.grouper
-            return cat.codes
+        if self._codes is not None:
+            # set in __init__
```
Comment (WillAyd) [inline pandas/core/groupby/grouper.py:592]: Hmm the comment here is a little confusing. So this is set in init but subsequently set to None?

Comment (jbrockmendel) [reply pandas/core/groupby/grouper.py:592]: self._codes is set in `__init__`, usually to None, but in one case to an ndarray, in which case we use that.  I'll flesh out the comment

**Comment by jreback** [comment]
sure, not sure if need to rebase as merged a couple of other refactors but prob can't hurt

**Comment by jbrockmendel** [comment]
rebased + green

**Pull request merged**